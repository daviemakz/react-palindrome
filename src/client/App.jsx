'use strict';

// Import NPM modules
import React, { Fragment, useState, useCallback, useEffect, memo } from 'react';
import PropTypes from 'prop-types';

// Load antDesign modules
import {
  Spin,
  Layout,
  Breadcrumb,
  Tag,
  Divider,
  Input,
  Button,
  Empty,
  message
} from 'antd';
const { Content, Footer } = Layout;
const BreadcrumbItem = Breadcrumb.Item;

// Set the top height to 5
export const configureMessages = () =>
  message.config({
    top: 5
  });

// Set the height of the messages
export const resetMessageConfig = () => {
  // Close all other ones
  message.destroy();
  // Configure a new instance
  configureMessages();
};

// Where the breadcrumb is built
export const BreadcrumbContainer = memo(() => {
  return (
    <Breadcrumb
      style={{
        margin: '16px 0'
      }}>
      <BreadcrumbItem id="level-1">Home</BreadcrumbItem>
      <BreadcrumbItem id="level-2">Palindrome Finder</BreadcrumbItem>
    </Breadcrumb>
  );
});

// Title of the page
export const Title = memo(() => (
  <h3
    style={{
      textAlign: 'left'
    }}>
    Please enter a string and click the button below:
  </h3>
));

// Controls which allows you to change the current palindrome
export const PageControls = memo(({ value, onChange, onSubmit, onRefresh }) => (
  <span style={{ display: 'flex' }}>
    <Input
      data-testid="input-enter-palindrome"
      value={value}
      onChange={e => onChange(e.target.value || '')}
      onPressEnter={onSubmit}
    />
    <Button
      data-testid="button-submit-palindrome"
      type="primary"
      onClick={onSubmit}
      style={{ marginLeft: '16px' }}>
      Test Palindrome
    </Button>
    <Button
      data-testid="button-refresh-palindrome-list"
      type="default"
      onClick={onRefresh}
      style={{ marginLeft: '16px' }}>
      Refresh List
    </Button>
  </span>
));

PageControls.propTypes = {
  value: PropTypes.string,
  onRefresh: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  onSubmit: PropTypes.func.isRequired
};

// Main area of the page
export const BodyArea = memo(({ palindromeList }) => (
  <Fragment>
    {palindromeList.length ? (
      palindromeList.map((number, index) => (
        <Tag style={{ marginBottom: '8px' }} key={index}>
          {number}
        </Tag>
      ))
    ) : (
      <Empty
        style={{ marginTop: '128px' }}
        description="No palindromes to display!"
      />
    )}
  </Fragment>
));

BodyArea.propTypes = {
  palindromeList: PropTypes.array
};

// Footer of the page
export const FooterArea = memo(() => (
  <Footer
    style={{
      textAlign: 'center'
    }}>
    NOTE: Any request sent to the backend will be proxied to an{' '}
    <a href="https://expressjs.com/" rel="noopener noreferrer" target="_blank">
      Express
    </a>{' '}
    server running on{' '}
    <a
      href="http://localhost:3000/palindromes"
      rel="noopener noreferrer"
      target="_blank">
      localhost:3000
    </a>{' '}
    via{' '}
    <a
      href="https://webpack.js.org/configuration/dev-server/"
      rel="noopener noreferrer"
      target="_blank">
      Webpack Dev Server
    </a>
    .
  </Footer>
));

// Export the main app
export default memo(() => {
  // Store selected palindrome and the overall list
  // Could have used useReducer here too...
  const [loading, updateLoading] = useState(false);
  const [palindrome, updatePalindrome] = useState('');
  const [palindromeList, updatePalindromeList] = useState([]);

  // HOOK: Save the reference for the refresh handler
  const handleOnRefresh = useCallback(() => {
    updateLoading(true);
    return fetch('api/palindromes')
      .then(result => result.json())
      .then(res => {
        updatePalindromeList(res.result);
        updateLoading(false);
      });
  }, []);

  // HOOK: Save the reference for the on submit handler
  const handleOnSubmit = useCallback(() => {
    const scopedPalindrome = palindrome;
    if (scopedPalindrome.length) {
      updateLoading(true);
      return fetch('api/palindromes', {
        method: 'POST',
        cache: 'no-cache',
        headers: {
          'Content-Type': 'application/json'
        },
        body: JSON.stringify({ data: scopedPalindrome })
      })
        .then(result => result.json())
        .then(res => {
          // Clear any messages which are on the screen
          resetMessageConfig();
          if (res.result) {
            // Show a notification
            message.success(
              `'${scopedPalindrome}' is a valid palindrome!`,
              1.5
            );
            // Refresh the list, skip changing loading for now as we are making another API call
            handleOnRefresh();
          } else {
            message.error(
              `'${scopedPalindrome}' is NOT a valid palindrome!`,
              1.5
            );
            // Lets just set loading to false and not fetch a list which hasnt changed
            updateLoading(false);
          }
        });
    }
    return message.warning(
      'Please enter some text in the field below before continuing...',
      1.5
    );
  }, [handleOnRefresh, palindrome]);

  // HOOK: Get the initial list of palindromes (only runs on initial render)
  useEffect(() => {
    handleOnRefresh();
  }, [handleOnRefresh]);

  // // Configure the message object
  useEffect(() => {
    configureMessages();
  }, []);

  // Return JSX
  return (
    <Layout className="layout">
      <Content style={{ padding: '0 50px' }}>
        <div>
          <BreadcrumbContainer />
        </div>
        <div
          style={{
            background: '#fff',
            padding: 0,
            paddingBottom: 24,
            minHeight: 280
          }}>
          <Spin spinning={loading}>
            <div
              style={{
                background: '#fff',
                padding: 24,
                minHeight: 'calc(100vh - 146px)'
              }}>
              <span>
                <Title />
                <PageControls
                  value={palindrome}
                  onChange={updatePalindrome}
                  onSubmit={handleOnSubmit}
                  onRefresh={handleOnRefresh}
                />
              </span>
              <Divider />
              <BodyArea palindromeList={palindromeList} />
            </div>
          </Spin>
        </div>
      </Content>
      <FooterArea />
    </Layout>
  );
});
