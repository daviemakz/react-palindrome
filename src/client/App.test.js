'use strict';

// Import NPM modules
import React from 'react';
import { render, fireEvent } from '@testing-library/react';
import { getByTestId } from '@testing-library/dom';

// Add JS-DOM and some few nifty assertions to JEST
import '@testing-library/jest-dom/extend-expect';

// Import app
import App, {
  configureMessages,
  resetMessageConfig,
  BreadcrumbContainer,
  Title,
  PageControls,
  BodyArea,
  FooterArea
} from './App';

// NOTE: We wont be doing too much simulation of click events while its very
// much possible to do. Lets leave that for more full spectrum integration
// testing.

/* eslint-disable-next-line */
// NOTE: Known issue (02/01/2020): https://github.com/testing-library/react-testing-library#suppressing-unnecessary-warnings-on-react-dom-168
const originalError = console.error;

// Test suite
describe('web application', () => {
  describe('functions', () => {
    describe('configureMessages()', () => {
      test('should not throw an error', () => {
        expect(configureMessages).not.toThrow();
      });
    });
    describe('resetMessageConfig()', () => {
      test('should not throw an error', () => {
        expect(resetMessageConfig).not.toThrow();
      });
    });
  });
  describe('react', () => {
    describe('<App/>', () => {
      beforeAll(() => {
        console.error = (...args) => {
          if (/Warning.*not wrapped in act/.test(args[0])) {
            return;
          }
          originalError.call(console, ...args);
        };
      });
      beforeEach(() => {
        // Reset the mocks
        fetch.resetMocks();
        // The next request will return the following data
        fetch.mockResponseOnce(
          JSON.stringify({ httpStatusCode: 200, result: ['aaa'] })
        );
      });
      test('renders without crashing', () => {
        const { container } = render(<App />);
        expect(container).toMatchSnapshot();
      });
      afterAll(() => {
        console.error = originalError;
      });
    });
    describe('<BreadcrumbContainer/>', () => {
      test('renders without crashing', () => {
        const { container } = render(<BreadcrumbContainer />);
        expect(container).toMatchSnapshot();
      });
    });
    describe('<Title/>', () => {
      test('renders without crashing', () => {
        const { container } = render(<Title />);
        expect(container).toMatchSnapshot();
      });
    });
    // NOTE: Due to writing this in a functional and modular way most of the interaction
    // apart from actual API calls done in this function we can test it independently
    // and pass any mocks / values down to the component
    describe('<PageControls/>', () => {
      const getDefaultProps = (propsToSpread = () => ({})) => ({
        value: '',
        onChange: () => void 0,
        onSubmit: () => void 0,
        onRefresh: () => void 0,
        ...propsToSpread
      });
      test('renders without crashing, input is empty', () => {
        const props = getDefaultProps({ value: '' });
        const { container } = render(<PageControls {...props} />);
        expect(container).toMatchSnapshot();
      });
      test('renders without crashing, input is NOT empty', () => {
        const props = getDefaultProps({ value: 'abcde fghij' });
        const { container } = render(<PageControls {...props} />);
        expect(container).toMatchSnapshot();
      });
      describe('event handlers', () => {
        // Store this ref to use later
        let container;
        // Build jest mocks
        const onChange = jest.fn(x => x);
        const onSubmit = jest.fn(x => x);
        const onRefresh = jest.fn(x => x);

        beforeEach(() => {
          // Render react component
          const props = getDefaultProps({ onChange, onSubmit, onRefresh });
          container = render(<PageControls {...props} />).container;
        });

        test('onChange() is called one with the correct value', () => {
          const testEnterPalindrome = getByTestId(
            container,
            'input-enter-palindrome'
          );
          fireEvent.change(testEnterPalindrome, {
            target: { value: 'abcdefg' }
          });
          expect(onChange.mock.calls.length).toBe(1);
          expect(onChange.mock.results[0].value).toBe('abcdefg');
        });

        test('onSubmit() is called when button is clicked', () => {
          const testPalindromeButton = getByTestId(
            container,
            'button-submit-palindrome'
          );
          fireEvent.click(testPalindromeButton);
          expect(onSubmit.mock.calls.length).toBe(1);
        });

        test('onRefresh() is called when button is clicked', () => {
          const testRefreshPalidromeList = getByTestId(
            container,
            'button-refresh-palindrome-list'
          );
          fireEvent.click(testRefreshPalidromeList);
          expect(onRefresh.mock.calls.length).toBe(1);
        });
      });
    });
    describe('<BodyArea/>', () => {
      test('renders with an empty list of palindromes', () => {
        const { container } = render(<BodyArea palindromeList={[]} />);
        expect(container).toMatchSnapshot();
      });
      test('renders with an non-empty list of palindromes', () => {
        const { container } = render(<BodyArea palindromeList={['otto']} />);
        expect(container).toMatchSnapshot();
      });
    });
    describe('<FooterArea/>', () => {
      test('renders without crashing', () => {
        const { container } = render(<FooterArea />);
        expect(container).toMatchSnapshot();
      });
    });
  });
});
