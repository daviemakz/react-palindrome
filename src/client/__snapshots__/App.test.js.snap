// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`web application react <App/> renders without crashing 1`] = `
<div>
  <section
    class="layout ant-layout"
  >
    <main
      class="ant-layout-content"
      style="padding: 0px 50px;"
    >
      <div>
        <div
          class="ant-breadcrumb"
          style="margin: 16px 0px;"
        >
          <span>
            <span
              class="ant-breadcrumb-link"
              id="level-1"
            >
              Home
            </span>
            <span
              class="ant-breadcrumb-separator"
            >
              /
            </span>
          </span>
          <span>
            <span
              class="ant-breadcrumb-link"
              id="level-2"
            >
              Palindrome Finder
            </span>
            <span
              class="ant-breadcrumb-separator"
            >
              /
            </span>
          </span>
        </div>
      </div>
      <div
        style="background: rgb(255, 255, 255); padding: 0px 0px 24px 0px; min-height: 280px;"
      >
        <div
          class="ant-spin-nested-loading"
        >
          <div>
            <div
              class="ant-spin ant-spin-spinning"
            >
              <span
                class="ant-spin-dot ant-spin-dot-spin"
              >
                <i
                  class="ant-spin-dot-item"
                />
                <i
                  class="ant-spin-dot-item"
                />
                <i
                  class="ant-spin-dot-item"
                />
                <i
                  class="ant-spin-dot-item"
                />
              </span>
            </div>
          </div>
          <div
            class="ant-spin-container ant-spin-blur"
          >
            <div
              style="background: rgb(255, 255, 255); padding: 24px; min-height: calc(100vh - 146px);"
            >
              <span>
                <h3
                  style="text-align: left;"
                >
                  Please enter a string and click the button below:
                </h3>
                <span
                  style="display: flex;"
                >
                  <input
                    class="ant-input"
                    data-testid="input-enter-palindrome"
                    type="text"
                    value=""
                  />
                  <button
                    class="ant-btn ant-btn-primary"
                    data-testid="button-submit-palindrome"
                    style="margin-left: 16px;"
                    type="button"
                  >
                    <span>
                      Test Palindrome
                    </span>
                  </button>
                  <button
                    class="ant-btn ant-btn-default"
                    data-testid="button-refresh-palindrome-list"
                    style="margin-left: 16px;"
                    type="button"
                  >
                    <span>
                      Refresh List
                    </span>
                  </button>
                </span>
              </span>
              <div
                class="ant-divider ant-divider-horizontal"
                role="separator"
              />
              <div
                class="ant-empty"
                style="margin-top: 128px;"
              >
                <div
                  class="ant-empty-image"
                >
                  <svg
                    height="152"
                    viewBox="0 0 184 152"
                    width="184"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <g
                      fill="none"
                      fill-rule="evenodd"
                    >
                      <g
                        transform="translate(24 31.67)"
                      >
                        <ellipse
                          cx="67.797"
                          cy="106.89"
                          fill="#F5F5F7"
                          fill-opacity=".8"
                          rx="67.797"
                          ry="12.668"
                        />
                        <path
                          d="M122.034 69.674L98.109 40.229c-1.148-1.386-2.826-2.225-4.593-2.225h-51.44c-1.766 0-3.444.839-4.592 2.225L13.56 69.674v15.383h108.475V69.674z"
                          fill="#AEB8C2"
                        />
                        <path
                          d="M101.537 86.214L80.63 61.102c-1.001-1.207-2.507-1.867-4.048-1.867H31.724c-1.54 0-3.047.66-4.048 1.867L6.769 86.214v13.792h94.768V86.214z"
                          fill="url(#linearGradient-1)"
                          transform="translate(13.56)"
                        />
                        <path
                          d="M33.83 0h67.933a4 4 0 0 1 4 4v93.344a4 4 0 0 1-4 4H33.83a4 4 0 0 1-4-4V4a4 4 0 0 1 4-4z"
                          fill="#F5F5F7"
                        />
                        <path
                          d="M42.678 9.953h50.237a2 2 0 0 1 2 2V36.91a2 2 0 0 1-2 2H42.678a2 2 0 0 1-2-2V11.953a2 2 0 0 1 2-2zM42.94 49.767h49.713a2.262 2.262 0 1 1 0 4.524H42.94a2.262 2.262 0 0 1 0-4.524zM42.94 61.53h49.713a2.262 2.262 0 1 1 0 4.525H42.94a2.262 2.262 0 0 1 0-4.525zM121.813 105.032c-.775 3.071-3.497 5.36-6.735 5.36H20.515c-3.238 0-5.96-2.29-6.734-5.36a7.309 7.309 0 0 1-.222-1.79V69.675h26.318c2.907 0 5.25 2.448 5.25 5.42v.04c0 2.971 2.37 5.37 5.277 5.37h34.785c2.907 0 5.277-2.421 5.277-5.393V75.1c0-2.972 2.343-5.426 5.25-5.426h26.318v33.569c0 .617-.077 1.216-.221 1.789z"
                          fill="#DCE0E6"
                        />
                      </g>
                      <path
                        d="M149.121 33.292l-6.83 2.65a1 1 0 0 1-1.317-1.23l1.937-6.207c-2.589-2.944-4.109-6.534-4.109-10.408C138.802 8.102 148.92 0 161.402 0 173.881 0 184 8.102 184 18.097c0 9.995-10.118 18.097-22.599 18.097-4.528 0-8.744-1.066-12.28-2.902z"
                        fill="#DCE0E6"
                      />
                      <g
                        fill="#FFF"
                        transform="translate(149.65 15.383)"
                      >
                        <ellipse
                          cx="20.654"
                          cy="3.167"
                          rx="2.849"
                          ry="2.815"
                        />
                        <path
                          d="M5.698 5.63H0L2.898.704zM9.259.704h4.985V5.63H9.259z"
                        />
                      </g>
                    </g>
                  </svg>
                </div>
                <p
                  class="ant-empty-description"
                >
                  No palindromes to display!
                </p>
              </div>
            </div>
          </div>
        </div>
      </div>
    </main>
    <footer
      class="ant-layout-footer"
      style="text-align: center;"
    >
      NOTE: Any request sent to the backend will be proxied to an
       
      <a
        href="https://expressjs.com/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Express
      </a>
       
      server running on
       
      <a
        href="http://localhost:3000/palindromes"
        rel="noopener noreferrer"
        target="_blank"
      >
        localhost:3000
      </a>
       
      via
       
      <a
        href="https://webpack.js.org/configuration/dev-server/"
        rel="noopener noreferrer"
        target="_blank"
      >
        Webpack Dev Server
      </a>
      .
    </footer>
  </section>
</div>
`;

exports[`web application react <BodyArea/> renders with an empty list of palindromes 1`] = `
<div>
  <div
    class="ant-empty"
    style="margin-top: 128px;"
  >
    <div
      class="ant-empty-image"
    >
      <svg
        height="152"
        viewBox="0 0 184 152"
        width="184"
        xmlns="http://www.w3.org/2000/svg"
      >
        <g
          fill="none"
          fill-rule="evenodd"
        >
          <g
            transform="translate(24 31.67)"
          >
            <ellipse
              cx="67.797"
              cy="106.89"
              fill="#F5F5F7"
              fill-opacity=".8"
              rx="67.797"
              ry="12.668"
            />
            <path
              d="M122.034 69.674L98.109 40.229c-1.148-1.386-2.826-2.225-4.593-2.225h-51.44c-1.766 0-3.444.839-4.592 2.225L13.56 69.674v15.383h108.475V69.674z"
              fill="#AEB8C2"
            />
            <path
              d="M101.537 86.214L80.63 61.102c-1.001-1.207-2.507-1.867-4.048-1.867H31.724c-1.54 0-3.047.66-4.048 1.867L6.769 86.214v13.792h94.768V86.214z"
              fill="url(#linearGradient-1)"
              transform="translate(13.56)"
            />
            <path
              d="M33.83 0h67.933a4 4 0 0 1 4 4v93.344a4 4 0 0 1-4 4H33.83a4 4 0 0 1-4-4V4a4 4 0 0 1 4-4z"
              fill="#F5F5F7"
            />
            <path
              d="M42.678 9.953h50.237a2 2 0 0 1 2 2V36.91a2 2 0 0 1-2 2H42.678a2 2 0 0 1-2-2V11.953a2 2 0 0 1 2-2zM42.94 49.767h49.713a2.262 2.262 0 1 1 0 4.524H42.94a2.262 2.262 0 0 1 0-4.524zM42.94 61.53h49.713a2.262 2.262 0 1 1 0 4.525H42.94a2.262 2.262 0 0 1 0-4.525zM121.813 105.032c-.775 3.071-3.497 5.36-6.735 5.36H20.515c-3.238 0-5.96-2.29-6.734-5.36a7.309 7.309 0 0 1-.222-1.79V69.675h26.318c2.907 0 5.25 2.448 5.25 5.42v.04c0 2.971 2.37 5.37 5.277 5.37h34.785c2.907 0 5.277-2.421 5.277-5.393V75.1c0-2.972 2.343-5.426 5.25-5.426h26.318v33.569c0 .617-.077 1.216-.221 1.789z"
              fill="#DCE0E6"
            />
          </g>
          <path
            d="M149.121 33.292l-6.83 2.65a1 1 0 0 1-1.317-1.23l1.937-6.207c-2.589-2.944-4.109-6.534-4.109-10.408C138.802 8.102 148.92 0 161.402 0 173.881 0 184 8.102 184 18.097c0 9.995-10.118 18.097-22.599 18.097-4.528 0-8.744-1.066-12.28-2.902z"
            fill="#DCE0E6"
          />
          <g
            fill="#FFF"
            transform="translate(149.65 15.383)"
          >
            <ellipse
              cx="20.654"
              cy="3.167"
              rx="2.849"
              ry="2.815"
            />
            <path
              d="M5.698 5.63H0L2.898.704zM9.259.704h4.985V5.63H9.259z"
            />
          </g>
        </g>
      </svg>
    </div>
    <p
      class="ant-empty-description"
    >
      No palindromes to display!
    </p>
  </div>
</div>
`;

exports[`web application react <BodyArea/> renders with an non-empty list of palindromes 1`] = `
<div>
  <span
    class="ant-tag"
    style="margin-bottom: 8px;"
  >
    otto
  </span>
</div>
`;

exports[`web application react <BreadcrumbContainer/> renders without crashing 1`] = `
<div>
  <div
    class="ant-breadcrumb"
    style="margin: 16px 0px;"
  >
    <span>
      <span
        class="ant-breadcrumb-link"
        id="level-1"
      >
        Home
      </span>
      <span
        class="ant-breadcrumb-separator"
      >
        /
      </span>
    </span>
    <span>
      <span
        class="ant-breadcrumb-link"
        id="level-2"
      >
        Palindrome Finder
      </span>
      <span
        class="ant-breadcrumb-separator"
      >
        /
      </span>
    </span>
  </div>
</div>
`;

exports[`web application react <FooterArea/> renders without crashing 1`] = `
<div>
  <footer
    class="ant-layout-footer"
    style="text-align: center;"
  >
    NOTE: Any request sent to the backend will be proxied to an
     
    <a
      href="https://expressjs.com/"
      rel="noopener noreferrer"
      target="_blank"
    >
      Express
    </a>
     
    server running on
     
    <a
      href="http://localhost:3000/palindromes"
      rel="noopener noreferrer"
      target="_blank"
    >
      localhost:3000
    </a>
     
    via
     
    <a
      href="https://webpack.js.org/configuration/dev-server/"
      rel="noopener noreferrer"
      target="_blank"
    >
      Webpack Dev Server
    </a>
    .
  </footer>
</div>
`;

exports[`web application react <PageControls/> renders without crashing, input is NOT empty 1`] = `
<div>
  <span
    style="display: flex;"
  >
    <input
      class="ant-input"
      data-testid="input-enter-palindrome"
      type="text"
      value="abcde fghij"
    />
    <button
      class="ant-btn ant-btn-primary"
      data-testid="button-submit-palindrome"
      style="margin-left: 16px;"
      type="button"
    >
      <span>
        Test Palindrome
      </span>
    </button>
    <button
      class="ant-btn ant-btn-default"
      data-testid="button-refresh-palindrome-list"
      style="margin-left: 16px;"
      type="button"
    >
      <span>
        Refresh List
      </span>
    </button>
  </span>
</div>
`;

exports[`web application react <PageControls/> renders without crashing, input is empty 1`] = `
<div>
  <span
    style="display: flex;"
  >
    <input
      class="ant-input"
      data-testid="input-enter-palindrome"
      type="text"
      value=""
    />
    <button
      class="ant-btn ant-btn-primary"
      data-testid="button-submit-palindrome"
      style="margin-left: 16px;"
      type="button"
    >
      <span>
        Test Palindrome
      </span>
    </button>
    <button
      class="ant-btn ant-btn-default"
      data-testid="button-refresh-palindrome-list"
      style="margin-left: 16px;"
      type="button"
    >
      <span>
        Refresh List
      </span>
    </button>
  </span>
</div>
`;

exports[`web application react <Title/> renders without crashing 1`] = `
<div>
  <h3
    style="text-align: left;"
  >
    Please enter a string and click the button below:
  </h3>
</div>
`;
